### PR TITLE
add scatter update feature

### PR DESCRIPTION
**Pull request by haoyouab**
Note:

- Race conditions are not handled in `tile_to_pxa` conversion. Leave it to accumulation/subgroup pass.

- According to the spec of [tf.tensor_scatter_nd_update](https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update):

> If indices contains duplicates, then their updates are accumulated (**SUMMED**).

And spec of [tf.scatter_update](https://www.tensorflow.org/api_docs/python/tf/compat/v1/scatter_update):

> If values in ref is to be updated more than once, because there are duplicate entries in indices, the order at which the updates happen for each value is **UNDEFINED**.

I choose to use the summing behavior as the default in `tile_to_pxa` conversion for now (i.e. all the elements/slices in the `updates` tensor under the same indices will be summed instead of racing/overwriting).

- ~`ScatterUpdateMode::SUM` is actually **NOT** updating anything (see [tf.scatter_add](https://www.tensorflow.org/api_docs/python/tf/compat/v1/scatter_add)). It basically uses scatterd slices and 'zero-like' slices to form a new tensor. Unlike the actual updating modes which **Creates a copy of the first input tensor with updated elements specified with second and third input tensors**. For now I just put it in `ScatterUpdateMode` as a default to simplify the interface. That's why it doesn't bother initialize the first input tensor `data` (i.e. it only needs the shape of data).~ Use `ScatterUpdateMode::NONE` to avoid confusion.

- ~Negative `axis` parameter is now proccessed in `tile_to_pxa` conversion. See [// TODO: Deal with negative axis earlier in ast?](https://github.com/Flex-plaidml-team/plaidml/pull/29/files#diff-1f30ed9e088aa5e1df6ea7e63e90239d2a3e9591ab86ab245d41316da15e4741R1369) in the source file.~ Deal with negative axis in edsl.h.

### CODE REVIEW

File: plaidml/edsl/tests/edsl_test.cc [hunk 1]
```diff
@@ -1658,42 +1658,194 @@ TEST_F(CppEdsl, Tan) {
...
   auto U = Placeholder(DType::FLOAT32, {4});
-  auto S = Placeholder(DType::INT32, {8});
-  auto O = scatter(U, I, S);
-  auto program = makeProgram("scatter", {I, U, S}, {O});
+  auto O = scatter(D, I, U);
+  auto program = makeProgram("scatter", {D, I, U}, {O});
 
+  // Don't bother initialize 'data' in SUM mode.
+  // Only the shape of data is needed.
+  std::vector<float> data;
```
Comment (YangleiZouIntel) [inline plaidml/edsl/tests/edsl_test.cc:1669]: Shouldn't data be initialized with some fixed length of 0?

File: plaidml/edsl/tests/edsl_test.cc [hunk 2]
```diff
@@ -1658,42 +1658,194 @@ TEST_F(CppEdsl, Tan) {
...
   std::vector<int32_t> indices = {4, 3, 1, 7};
```
Comment (YangleiZouIntel) [inline plaidml/edsl/tests/edsl_test.cc:1670]: Is this testing indices with summing of update values with same indices? shouldn't indices contain same values like {4,3,3,7}?

Comment (haoyouab) [reply plaidml/edsl/tests/edsl_test.cc:1670]: Yes. Will add more test cases with duplicate indices later.

File: plaidml/edsl/tests/edsl_test.cc [hunk 3]
```diff
@@ -1659,41 +1659,237 @@ TEST_F(CppEdsl, Tan) {
...
+  std::vector<float> expected = {
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  //
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  //
+      10, 10, 10, 10, 12, 12, 12, 12, 14, 14, 14, 14, 16, 16, 16, 16,  //
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0   //
+  };
+  checkExact(program, {data, indices, updates}, {expected});
+}
+
+TEST_F(CppEdsl, Scatter3DUpdateSlice) {
```
Comment (YangleiZouIntel) [inline plaidml/edsl/tests/edsl_test.cc:1757]: Please rename to ScatterNDUpdateSlice

File: plaidml/edsl/tests/edsl_test.cc [hunk 4]
```diff
@@ -1659,41 +1659,237 @@ TEST_F(CppEdsl, Tan) {
...
+  std::vector<float> expected = {
+      9, 1, 9, 1,
+      1, 9, 1, 9,
+      1, 1, 1, 1,
+      1, 1, 1, 1
+  };
+  checkExact(program, {data, indices, updates}, {expected});
+};
+
+TEST_F(CppEdsl, ScatterNDUpdateSlice3) {
```
Comment (YangleiZouIntel) [inline plaidml/edsl/tests/edsl_test.cc:1831]: Please sort all scatter tests by name

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 5]
```diff
@@ -1362,35 +1364,111 @@ struct ScatterOpConversion : public OpConversionPattern<tile::ScatterOp> {
 
     // Load the location value from the indices tensor.
     // Create an affine map for loading the index, using leading counters.
-    size_t idxDims = indices.getType().cast<MemRefType>().getShape().size();
+    int axis = *(op.axis().getRawData());
```
Comment (YangleiZouIntel) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: The convention is defining axis as an IndexAttr, which is signless. To avoid confusion, it might be better to  deal negative axis earlier. In gather, currently negative axis is converted  as soon as passing to the gather class.

File: plaidml/edsl/tests/edsl_test.cc (hunk 3 above)
Comment (YangleiZouIntel) [reply plaidml/edsl/tests/edsl_test.cc:1757]: My mistake.. I think there is another test with the same name... please keep it as Scatter3D..

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc (hunk 5 above)
Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: Got it. Will deal with negative number in the way as gather does.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 6]
```diff
@@ -1362,35 +1364,111 @@ struct ScatterOpConversion : public OpConversionPattern<tile::ScatterOp> {
...
+    if (axis < 0) {
+      axis = dstDims + axis;
+    }
+
+    auto idxShape = indices.getType().cast<MemRefType>().getShape();
+    size_t idxDims = idxShape.size();
     auto idxLoadMap = AffineMap::getMultiDimIdentityMap(idxDims, ctx);
-    auto idxLoadOps = loop.getIVs().take_front(idxDims);
+    SmallVector<Value, 4> dstOps;
+    // TODO: Move this code block into separate functions?
```
Comment (YangleiZouIntel) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: Yes please move the if else block to a new function, and use 
`if(op.updateMode() == ScatterUpdateMode::nd)) {}else{}`
instead of 
`if(op.updateMode() != ScatterUpdateMode::nd)) {}else{}`

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 7]
```diff
@@ -1362,35 +1364,107 @@ struct ScatterOpConversion : public OpConversionPattern<tile::ScatterOp> {
...
+        dstOps.push_back(loop.getIVs()[i]);
+      }
     }
 
+    if (op.updateMode() != ScatterUpdateMode::none) {
+      auto elementType = data.getType().cast<MemRefType>().getElementType();
+      auto zeroVal = rewriter.create<mlir::ConstantOp>(
+          loc, elementType, rewriter.getFloatAttr(elementType, 0.0));
+
+      rewriter.create<mlir::StoreOp>(loc, zeroVal, data, dstOps);
```
Comment (YangleiZouIntel) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: This changes the input data... We should keep the input data unchanged, shouldn't we?

Comment (YangleiZouIntel) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: Another thing is should Update mode always assume indices has no duplicates? Maybe for update mode we should just override for duplicate indices instead of sum.

Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: Yes. Maybe just copying the input to the output is enough.

Comment (haoyouab) [reply pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: > Another thing is should Update mode always assume indices has no duplicates? Maybe for update mode we should just override for duplicate indices instead of sum.

Yes but the problem is that updating can be happening on multiple threads (if this is in a parallel op or something) so which one will be overwrriten is **UNDEFINED**.

File: pmlc/ast/ast_ops.cc [hunk 8]
```diff
@@ -181,12 +182,24 @@ struct ScatterOp : Intrinsic {
...
           "'scatter' requires operand #2 to be an integer.");
     }
-    // shape tensor
-    TensorShape ret(shapes[0].elementType);
-    ret.sizes.push_back(shapes[2].sizes[0]);
-    for (size_t i = shapes[1].getRank(); i < shapes[0].getRank(); i++) {
-      ret.sizes.push_back(shapes[0].sizes[i]);
+
+    auto axis = getIntegerValue(evaluator, operands[3]);
+    // if (!axis || axis.getValue() >= rank || axis.getValue() < 0) {
```
Comment (YangleiZouIntel) [inline pmlc/ast/ast_ops.cc]: Please remove this.

**Comment by YangleiZouIntel** [comment]
Please fix  [scatter mlir tests](https://github.com/Flex-plaidml-team/plaidml/blob/zhibinli-scatter-feature/pmlc/conversion/tile_to_pxa/tests/scatter.mlir), which fails on this PR.
You can run `bazelisk test //...` to check all tests, only 4 tests under //pmlc/rt/opencl/ and  //pmlc/rt/vulkan/ are allowed to fail, all other tests  must pass.
You can also use this to check this single test.
`bazelisk test //pmlc/conversion/tile_to_pxa/tests:scatter.mlir.test`

**Comment by haoyouab** [comment]
> Please fix [scatter mlir tests](https://github.com/Flex-plaidml-team/plaidml/blob/zhibinli-scatter-feature/pmlc/conversion/tile_to_pxa/tests/scatter.mlir), which fails on this PR.
> You can run `bazelisk test //...` to check all tests, only 4 tests under //pmlc/rt/opencl/ and //pmlc/rt/vulkan/ are allowed to fail, all other tests must pass.
> You can also use this to check this single test.
> `bazelisk test //pmlc/conversion/tile_to_pxa/tests:scatter.mlir.test`

I temporarily fix the error of scatter.mlir but haven't added more tests regarding the new feature. Will do that later.

**Comment by haoyouab** [comment]
> Please fix [scatter mlir tests](https://github.com/Flex-plaidml-team/plaidml/blob/zhibinli-scatter-feature/pmlc/conversion/tile_to_pxa/tests/scatter.mlir), which fails on this PR.
> You can run `bazelisk test //...` to check all tests, only 4 tests under //pmlc/rt/opencl/ and //pmlc/rt/vulkan/ are allowed to fail, all other tests must pass.
> You can also use this to check this single test.
> `bazelisk test //pmlc/conversion/tile_to_pxa/tests:scatter.mlir.test`

Is this about FileCheck? Which specific version is required?

**Comment by haoyouab** [comment]
Merged in upstream https://github.com/plaidml/plaidml/pull/1610.
